source/unified-test-format/unified-test-format.rst

```diff
@@ -0,0 +1,1896 @@
+
+.. _commonOptions_readConcern:
+
+- ``readConcern``: Optional document. Map of parameters to construct a read
+  concern.
+
+  The structure of this document is as follows:
+
+  - ``level``: Required string.
+
+.. _commonOptions_readPreference:
+
+- ``readPreference``: Optional document. Map of parameters to construct a read
+  preference.
+
+  The structure of this document is as follows:
+
+  - ``mode``: Required string.
+
+  - ``tagSets``: Optional array of documents.
+
+  - ``maxStalenessSeconds``: Optional integer.
+
+  - ``hedge``: Optional document.
+
+.. _commonOptions_session:
+
+- ``session``: Optional string. Session entity which will be resolved to a
+  ClientSession object. The YAML file SHOULD use an `alias node`_ for a session
+  entity's ``id`` field (e.g. ``session: *session0``).
+
+.. _commonOptions_writeConcern:
+
+- ``writeConcern``: Optional document. Map of parameters to construct a write
+  concern.
+
+  The structure of this document is as follows:
+
+  - ``journal``: Optional boolean.
+
+  - ``w``: Optional integer or string.
+
+  - ``wtimeoutMS``: Optional integer.
+
+
+Version String
+--------------
+
+Version strings, which are used for `schemaVersion`_ and `runOn`_, MUST conform
+to one of the following formats, where each component is an integer:
+
+- ``<major>.<minor>.<patch>``
+- ``<major>.<minor>`` (``<patch>`` is assumed to be zero)
+- ``<major>`` (``<minor>`` and ``<patch>`` are assumed to be zero)
+
+
+Entity Test Operations
+----------------------
+
+Entity operations correspond to an API method on a driver object. If
+`operation.object <operation_object_>`_ refers to an `entity`_ name (e.g.
+"collection0") then `operation.name <operation_name_>`_ is expected to reference
+an API method on that class.
+
+Some specifications group optional parameters for API methods under an
+``options`` parameter (e.g. ``options: Optional<UpdateOptions>`` in the CRUD
+spec). While this improves readability of the spec document(s) by allowing
+option documentation to be expanded and reused, it would add unnecessary
+verbosity to test files. Therefore, test files SHALL declare all required and
+optional parameters for an API method directly within
+`operation.arguments <operation_arguments_>`_ (e.g. ``upsert`` for ``updateOne``
+is *not* nested under an ``options`` key), unless otherwise stated below.
+
+If ``session`` is specified in ``operation.arguments`_, it is defined according
+to `commonOptions_session`_. Test runners MUST resolve the ``session`` argument
+to session entity *before* passing it as a parameter to any API method.
+
+This spec does not provide exhaustive documentation for all possible API methods
+that may appear in a test; however, the following sections discuss all supported
+entities and their operations in some level of detail. Special handling for
+certain operations is also discussed below.
+
+**TODO**: While CRUD methods tend to flatten options into ``arguments``
```

Comment:
This note was leftover from the previous review and I probably should have removed it in the last push.

I don't see a problem with flattening options into `arguments`, as that's already done today. I originally had much longer explanation of why specs define `options` parameters, but ultimately shortened it to:

> While this (`options` parameters) improves readability of the spec document(s) by allowing option documentation to be expanded and reused, it would add unnecessary verbosity to test files.

Without model classes like UpdateOptions, the API declaration for each update method in the CRUD spec would run off the page (assuming no word wrap in a code block). This also lets the spec re-use the options across several methods (notably not possible in the bulk write models). My understanding is that this convention is common in specs today because it comes back to what was originally done in CRUD, and that was purely for stylistic reasons. Consider that the bulk write models (e.g. UpdateOneModel) defines its options alongside its required parameters (no `options` nesting).

> Given a flattened option map, we can't construct the hierarchical option map so it seems like we have to go with specifying everything under 'options'.

Consider:

```yaml
name: find
arguments:
  filter: { }
  allowDiskUse: false
```

I don't think any driver has a problem processing that today. In PHP, I extract the required/positional arguments from the `arguments` document and whatever is left is passed as options. In Python, I assume they're both named parameters and you just pass them as-is. Presumably, this isn't an issue for typed languages either.

Am I misunderstanding something?